src/shared/bpf-program.c

```diff
@@ -11,6 +11,30 @@
 #include "memory-util.h"
 #include "missing_syscall.h"
 #include "path-util.h"
+#include "string-table.h"
+
+static const char *const bpf_cgroup_attach_type_table[__MAX_BPF_ATTACH_TYPE] = {
+        [BPF_CGROUP_INET_INGRESS] = "ingress",
+        [BPF_CGROUP_INET_EGRESS] = "egress",
+        [BPF_CGROUP_INET_SOCK_CREATE] = "sock_create",
+        [BPF_CGROUP_SOCK_OPS] = "sock_ops",
+        [BPF_CGROUP_DEVICE] = "device",
+        [BPF_CGROUP_INET4_BIND] = "bind4",
+        [BPF_CGROUP_INET6_BIND] = "bind6",
+        [BPF_CGROUP_INET4_CONNECT] = "connect4",
+        [BPF_CGROUP_INET6_CONNECT] = "connect6",
+        [BPF_CGROUP_INET4_POST_BIND] = "post_bind4",
+        [BPF_CGROUP_INET6_POST_BIND] = "post_bind6",
+        [BPF_CGROUP_UDP4_SENDMSG] = "sendmsg4",
+        [BPF_CGROUP_UDP6_SENDMSG] = "sendmsg6",
+        [BPF_CGROUP_SYSCTL] = "sysctl",
+        [BPF_CGROUP_UDP4_RECVMSG] = "recvmsg4",
+        [BPF_CGROUP_UDP6_RECVMSG] = "recvmsg6",
+        [BPF_CGROUP_GETSOCKOPT] = "getsockopt",
+        [BPF_CGROUP_SETSOCKOPT] = "setsockopt",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(bpf_cgroup_attach_type, enum bpf_attach_type);
```

Comment:
hmm, bpf_attach_type is an enum that only has positive values. That unfortunately is incompatible with DEFINE_STRING_TABLE_LOOKUP() though, as the functions that this macro generates will return -1 on failure, and an enum with only positive values likely doesn't know a negative value (as the compiler is free to pick any type it likes, including positive ones). 

A simple fix is to just specify "int" as type here, and use the macro anyway. You lose a bit of typesafety, but should be ok...

(this then means that you need to assign the bpf_cgroup_attach_type_from_string() return value to an `int` type first, check for negativity, then cast to the enum type for further use)

You'll find that all our own enums always carry a special enum value called `_xyz_INVALID` that is set to -1, to avoid this problem and making sure -1 is always a valid return value. But for foreign defined enums this unfortunately doesn#t work...

Reply:
switched to `int`.